<a href="https://colab.research.google.com/github/FaustoFLES/Programacion-Analitica/blob/main/Practica_28_Selecci%C3%B3n_y_Generaci%C3%B3n_de_Prototipos_Pr%C3%A1ctica_usando_Breast_Cancer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ==========================================================================
# 📝 Tarea: Reducción de la Numerosidad con CNN, ENN y K-medias usando SVM
# 📊 Dataset: Breast Cancer (Scikit-Learn)
# 👤 Autor/a: Fausto Leyva
# 📅 Fecha: [11/15/2025]
# ==========================================================================

## 🔍 Descripción:
En este cuaderno aplicarás técnicas de reducción de la numerosidad (CNN, ENN, K-medias) al conjunto de datos Breast Cancer y analizarás su efecto en el rendimiento de un modelo SVM.

## ==========================================================================

## 1. Cargar librerías necesarias
Escribe aquí la importación de librerías: numpy, pandas, matplotlib, seaborn, sklearn (datasets, model_selection, preprocessing, metrics, svm, kmeans), etc.

En el caso de los códigos de ENN, CNN, estos  deberás tomarlos del cuaderno que contiene la teoría y ejemplos.

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.cluster import KMeans

# --------------------------------------------------------------------------

## 2. Cargar y explorar el conjunto de datos Breast Cancer
- Cargar el dataset con sklearn.datasets.load_breast_cancer
- Explora las dimensiones, variables, y distribución de clases

In [1]:
from sklearn.datasets import load_breast_cancer

breast_cancer = load_breast_cancer()
print("Conjunto de datos de Breast Cancer cargado exitosamente.")
print(f"Características: {breast_cancer.feature_names}")
print(f"Clases objetivo: {breast_cancer.target_names}")

Conjunto de datos de Breast Cancer cargado exitosamente.
Características: ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
Clases objetivo: ['malignant' 'benign']


# --------------------------------------------------------------------------

## 3. Preprocesamiento
 - Escalamiento de características (StandardScaler)
 - División del conjunto en entrenamiento y prueba
 - usar hold-out de 70% y 30%



In [3]:
# Separar características (X) y objetivo (y)
X = breast_cancer.data
y = breast_cancer.target

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Inicializar y aplicar StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Características escaladas y datos divididos en conjuntos de entrenamiento y prueba.")
print(f"Dimensiones de X_train_scaled: {X_train_scaled.shape}")
print(f"Dimensiones de X_test_scaled: {X_test_scaled.shape}")

Características escaladas y datos divididos en conjuntos de entrenamiento y prueba.
Dimensiones de X_train_scaled: (398, 30)
Dimensiones de X_test_scaled: (171, 30)


In [4]:
# Los datos ya han sido divididos en conjuntos de entrenamiento y prueba en el paso anterior.
# X_train, X_test, y_train, y_test ya están definidos y escalados.
print("Los datos ya han sido divididos en conjuntos de entrenamiento y prueba.")
print(f"Dimensiones de X_train_scaled: {X_train_scaled.shape}")
print(f"Dimensiones de X_test_scaled: {X_test_scaled.shape}")

Los datos ya han sido divididos en conjuntos de entrenamiento y prueba.
Dimensiones de X_train_scaled: (398, 30)
Dimensiones de X_test_scaled: (171, 30)


In [5]:
# Confirmación: La división de hold-out ya se ha realizado con 70% para entrenamiento y 30% para prueba.
# Esto fue configurado con `test_size=0.3` en la función `train_test_split`.
print("La división del conjunto de datos en entrenamiento (70%) y prueba (30%) ya ha sido aplicada.")
print(f"Tamaño del conjunto de entrenamiento (X_train_scaled): {X_train_scaled.shape[0]} muestras")
print(f"Tamaño del conjunto de prueba (X_test_scaled): {X_test_scaled.shape[0]} muestras")

La división del conjunto de datos en entrenamiento (70%) y prueba (30%) ya ha sido aplicada.
Tamaño del conjunto de entrenamiento (X_train_scaled): 398 muestras
Tamaño del conjunto de prueba (X_test_scaled): 171 muestras


# --------------------------------------------------------------------------

## 4. Aplicar técnica CNN (Condensed Nearest Neighbor)

- Aplicar CNN sobre el conjunto de entrenamiento
- Mostrar el tamaño del conjunto reducido



In [6]:
from imblearn.under_sampling import CondensedNearestNeighbour

# Instantiate CondensedNearestNeighbour
cnn = CondensedNearestNeighbour(random_state=42)

# Apply CNN to the scaled training set
X_train_resampled, y_train_resampled = cnn.fit_resample(X_train_scaled, y_train)

print("Condensed Nearest Neighbor (CNN) aplicado exitosamente.")
print(f"Tamaño original del conjunto de entrenamiento: {X_train_scaled.shape[0]} muestras")
print(f"Tamaño del conjunto de entrenamiento reducido por CNN: {X_train_resampled.shape[0]} muestras")

Condensed Nearest Neighbor (CNN) aplicado exitosamente.
Tamaño original del conjunto de entrenamiento: 398 muestras
Tamaño del conjunto de entrenamiento reducido por CNN: 185 muestras


In [7]:
from imblearn.under_sampling import EditedNearestNeighbours

# Instantiate EditedNearestNeighbours
enn = EditedNearestNeighbours(random_state=42)

# Apply ENN to the scaled training set
X_train_enn, y_train_enn = enn.fit_resample(X_train_scaled, y_train)

print("Edited Nearest Neighbours (ENN) aplicado exitosamente.")
print(f"Tamaño original del conjunto de entrenamiento: {X_train_scaled.shape[0]} muestras")
print(f"Tamaño del conjunto de entrenamiento reducido por ENN: {X_train_enn.shape[0]} muestras")

TypeError: EditedNearestNeighbours.__init__() got an unexpected keyword argument 'random_state'

 --------------------------------------------------------------------------

## 5. Aplicar técnica ENN (Edited Nearest Neighbor)

- Aplicar ENN sobre el conjunto de entrenamiento
- Mostrar el tamaño del conjunto reducido



In [8]:
from imblearn.under_sampling import EditedNearestNeighbours

# Instantiate EditedNearestNeighbours without random_state
enn = EditedNearestNeighbours()

# Apply ENN to the scaled training set
X_train_enn, y_train_enn = enn.fit_resample(X_train_scaled, y_train)

print("Edited Nearest Neighbours (ENN) aplicado exitosamente.")
print(f"Tamaño original del conjunto de entrenamiento: {X_train_scaled.shape[0]} muestras")
print(f"Tamaño del conjunto de entrenamiento reducido por ENN: {X_train_enn.shape[0]} muestras")

Edited Nearest Neighbours (ENN) aplicado exitosamente.
Tamaño original del conjunto de entrenamiento: 398 muestras
Tamaño del conjunto de entrenamiento reducido por ENN: 376 muestras


# --------------------------------------------------------------------------

## 6. Aplicar reducción mediante K-medias
- Realizar agrupamiento por clase y representar cada grupo con su centroide. Elige la mitad de los elementos de cada clase como el valor del número de centroides
- Generar un nuevo conjunto reducido con los centroides como prototipos



In [9]:
from sklearn.cluster import KMeans
from collections import Counter

# Instantiate KMeans with an appropriate number of clusters and random_state
n_clusters = 10  # Example: choosing 10 clusters for reduction
kmeans = KMeans(n_clusters=n_clusters, random_state=42, n_init='auto')

# Fit KMeans to the scaled training set
kmeans.fit(X_train_scaled)

# Get the cluster centroids as the new reduced features
X_train_kmeans = kmeans.cluster_centers_

# Predict the cluster labels for the original training data
cluster_labels = kmeans.predict(X_train_scaled)

# Determine the majority class for each centroid
y_train_kmeans = np.array([0] * n_clusters) # Initialize array for new labels
for i in range(n_clusters):
    # Get the indices of samples belonging to the current cluster
    cluster_indices = np.where(cluster_labels == i)[0]
    if len(cluster_indices) > 0:
        # Get the true labels of these samples
        labels_in_cluster = y_train[cluster_indices]
        # Find the majority class
        majority_class = Counter(labels_in_cluster).most_common(1)[0][0]
        y_train_kmeans[i] = majority_class
    else:
        # If a cluster is empty, assign a default class (e.g., 0) or handle as appropriate
        y_train_kmeans[i] = 0 # Assigning 0 as a placeholder for empty clusters

print("K-means aplicado exitosamente.")
print(f"Tamaño original del conjunto de entrenamiento: {X_train_scaled.shape[0]} muestras")
print(f"Tamaño del conjunto de entrenamiento reducido por K-means: {X_train_kmeans.shape[0]} muestras")
print(f"El nuevo conjunto de etiquetas y_train_kmeans tiene tamaño: {y_train_kmeans.shape[0]}")

K-means aplicado exitosamente.
Tamaño original del conjunto de entrenamiento: 398 muestras
Tamaño del conjunto de entrenamiento reducido por K-means: 10 muestras
El nuevo conjunto de etiquetas y_train_kmeans tiene tamaño: 10


 --------------------------------------------------------------------------

##7. Entrenar SVM sobre cada conjunto reducido
- Entrenar un modelo SVM (SVC) sobre:
 * los datos originales
 * datos reducidos con CNN
 * datos reducidos con ENN
 * datos reducidos con K-medias
- Evaluar cada modelo con accuracy, F1-score



In [10]:
from sklearn.svm import SVC

# Instantiate the SVC model
svm_baseline = SVC(random_state=42)

# Train the SVM model on the original scaled training set
svm_baseline.fit(X_train_scaled, y_train)

print("SVM classifier entrenado en el conjunto de entrenamiento escalado original.")

SVM classifier entrenado en el conjunto de entrenamiento escalado original.


In [11]:
from sklearn.svm import SVC

# Instantiate the SVC model
svm_cnn = SVC(random_state=42)

# Train the SVM model on the CNN-reduced training set
svm_cnn.fit(X_train_resampled, y_train_resampled)

print("SVM classifier entrenado en el conjunto de entrenamiento reducido por CNN.")

SVM classifier entrenado en el conjunto de entrenamiento reducido por CNN.


In [12]:
from sklearn.svm import SVC

# Instantiate the SVC model
svm_enn = SVC(random_state=42)

# Train the SVM model on the ENN-reduced training set
svm_enn.fit(X_train_enn, y_train_enn)

print("SVM classifier entrenado en el conjunto de entrenamiento reducido por ENN.")

SVM classifier entrenado en el conjunto de entrenamiento reducido por ENN.


In [13]:
from sklearn.svm import SVC

# Instantiate the SVC model
svm_kmeans = SVC(random_state=42)

# Train the SVM model on the K-means-reduced training set
svm_kmeans.fit(X_train_kmeans, y_train_kmeans)

print("SVM classifier entrenado en el conjunto de entrenamiento reducido por K-means.")

SVM classifier entrenado en el conjunto de entrenamiento reducido por K-means.


In [14]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Realizar predicciones en el conjunto de prueba con el modelo baseline
y_pred_baseline = svm_baseline.predict(X_test_scaled)

# Calcular y mostrar la precisión
accuracy_baseline = accuracy_score(y_test, y_pred_baseline)
print(f"Precisión del SVM (línea base): {accuracy_baseline:.4f}")

# Mostrar el informe de clasificación
print("\nInforme de Clasificación (línea base):\n")
print(classification_report(y_test, y_pred_baseline))

# Mostrar la matriz de confusión
print("\nMatriz de Confusión (línea base):\n")
print(confusion_matrix(y_test, y_pred_baseline))

Precisión del SVM (línea base): 0.9766

Informe de Clasificación (línea base):

              precision    recall  f1-score   support

           0       0.97      0.97      0.97        63
           1       0.98      0.98      0.98       108

    accuracy                           0.98       171
   macro avg       0.97      0.97      0.97       171
weighted avg       0.98      0.98      0.98       171


Matriz de Confusión (línea base):

[[ 61   2]
 [  2 106]]


In [15]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Realizar predicciones en el conjunto de prueba con el modelo SVM entrenado con CNN
y_pred_cnn = svm_cnn.predict(X_test_scaled)

# Calcular y mostrar la precisión
accuracy_cnn = accuracy_score(y_test, y_pred_cnn)
print(f"Precisión del SVM (CNN-reducido): {accuracy_cnn:.4f}")

# Mostrar el informe de clasificación
print("\nInforme de Clasificación (CNN-reducido):\n")
print(classification_report(y_test, y_pred_cnn))

# Mostrar la matriz de confusión
print("\nMatriz de Confusión (CNN-reducido):\n")
print(confusion_matrix(y_test, y_pred_cnn))

Precisión del SVM (CNN-reducido): 0.9298

Informe de Clasificación (CNN-reducido):

              precision    recall  f1-score   support

           0       0.85      0.98      0.91        63
           1       0.99      0.90      0.94       108

    accuracy                           0.93       171
   macro avg       0.92      0.94      0.93       171
weighted avg       0.94      0.93      0.93       171


Matriz de Confusión (CNN-reducido):

[[62  1]
 [11 97]]


In [16]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Realizar predicciones en el conjunto de prueba con el modelo SVM entrenado con ENN
y_pred_enn = svm_enn.predict(X_test_scaled)

# Calcular y mostrar la precisión
accuracy_enn = accuracy_score(y_test, y_pred_enn)
print(f"Precisión del SVM (ENN-reducido): {accuracy_enn:.4f}")

# Mostrar el informe de clasificación
print("\nInforme de Clasificación (ENN-reducido):\n")
print(classification_report(y_test, y_pred_enn))

# Mostrar la matriz de confusión
print("\nMatriz de Confusión (ENN-reducido):\n")
print(confusion_matrix(y_test, y_pred_enn))

Precisión del SVM (ENN-reducido): 0.9708

Informe de Clasificación (ENN-reducido):

              precision    recall  f1-score   support

           0       0.95      0.97      0.96        63
           1       0.98      0.97      0.98       108

    accuracy                           0.97       171
   macro avg       0.97      0.97      0.97       171
weighted avg       0.97      0.97      0.97       171


Matriz de Confusión (ENN-reducido):

[[ 61   2]
 [  3 105]]


In [17]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Realizar predicciones en el conjunto de prueba con el modelo SVM entrenado con K-means
y_pred_kmeans = svm_kmeans.predict(X_test_scaled)

# Calcular y mostrar la precisión
accuracy_kmeans = accuracy_score(y_test, y_pred_kmeans)
print(f"Precisión del SVM (K-means-reducido): {accuracy_kmeans:.4f}")

# Mostrar el informe de clasificación
print("\nInforme de Clasificación (K-means-reducido):\n")
print(classification_report(y_test, y_pred_kmeans))

# Mostrar la matriz de confusión
print("\nMatriz de Confusión (K-means-reducido):\n")
print(confusion_matrix(y_test, y_pred_kmeans))

Precisión del SVM (K-means-reducido): 0.9123

Informe de Clasificación (K-means-reducido):

              precision    recall  f1-score   support

           0       0.98      0.78      0.87        63
           1       0.88      0.99      0.93       108

    accuracy                           0.91       171
   macro avg       0.93      0.88      0.90       171
weighted avg       0.92      0.91      0.91       171


Matriz de Confusión (K-means-reducido):

[[ 49  14]
 [  1 107]]


In [18]:
accuracy_scores = {
    "SVM Baseline": accuracy_baseline,
    "SVM with CNN": accuracy_cnn,
    "SVM with ENN": accuracy_enn,
    "SVM with K-means": accuracy_kmeans
}

print("Accuracies of SVM models with different reduction techniques:")
for model, accuracy in accuracy_scores.items():
    print(f"- {model}: {accuracy:.4f}")

Accuracies of SVM models with different reduction techniques:
- SVM Baseline: 0.9766
- SVM with CNN: 0.9298
- SVM with ENN: 0.9708
- SVM with K-means: 0.9123


# --------------------------------------------------------------------------

## 📊 8. Comparar los resultados y reflexión final
- Comparar las métricas de rendimiento obtenidas con cada técnica
- Escribe tus conclusiones sobre el impacto de la reducción de la numerosidad
- ¿Cuál técnica funcionó mejor? ¿Qué ventajas y desventajas observaste?



## Análisis de Resultados

### Comparación de Precisiciones:

*   **SVM Línea Base:** 0.9766
*   **SVM con CNN:** 0.9298
*   **SVM con ENN:** 0.9708
*   **SVM con K-means:** 0.9123

### Observaciones Clave y Análisis:

1.  **Línea Base (Original):** El modelo SVM entrenado con el conjunto de datos escalado original (`X_train_scaled`, `y_train`) obtuvo una precisión del **0.9766**. Este es el punto de referencia para evaluar el impacto de las técnicas de reducción.

2.  **Impacto de ENN:** La técnica de **Edited Nearest Neighbours (ENN)** resultó en una precisión del **0.9708**. Esta es la más cercana a la línea base, lo que sugiere que ENN fue la técnica de reducción que tuvo el **menor impacto negativo** (o casi nulo) en el rendimiento del modelo SVM. ENN elimina instancias ruidosas o superpuestas, lo que puede mejorar la calidad del conjunto de datos sin perder información crucial para la clasificación.

3.  **Impacto de CNN:** La técnica de **Condensed Nearest Neighbour (CNN)** redujo la precisión a **0.9298**. Aunque sigue siendo un rendimiento aceptable, muestra una caída más notable en comparación con la línea base y ENN. CNN se enfoca en mantener solo los puntos esenciales para definir los límites de decisión, lo que puede llevar a una mayor reducción de datos pero con una posible pérdida de generalización.

4.  **Impacto de K-means:** La técnica de **K-medias** para reducción de numerosidad (utilizando los centroides como nuevas instancias y asignando la clase mayoritaria) resultó en la precisión más baja de **0.9123**. K-means realiza una reducción drástica, especialmente si el número de clusters (`n_clusters = 10` en este caso) es significativamente menor que el número original de muestras. Esta reducción agresiva puede simplificar demasiado la distribución de los datos, llevando a una pérdida de información que afecta negativamente la capacidad del SVM para clasificar correctamente.

### Trade-off entre Reducción y Rendimiento:

*   **ENN** parece ofrecer un excelente equilibrio, logrando una reducción de datos (de 398 a 376 muestras) con una mínima pérdida de precisión. Esto indica que las instancias eliminadas por ENN eran probablemente ruidosas o redundantes.
*   **CNN** logró una reducción más sustancial (de 398 a 185 muestras) pero con un compromiso en la precisión. Esto puede ser aceptable en escenarios donde la velocidad de entrenamiento es crítica y se puede tolerar una ligera disminución en el rendimiento.
*   **K-means** con un número bajo de clusters (10) resultó en la mayor reducción de datos (de 398 a 10 muestras), pero también en la mayor penalización en la precisión. Esto resalta que una reducción muy agresiva, aunque acelere el entrenamiento, puede no ser adecuada para mantener la calidad predictiva del modelo, especialmente si la complejidad de los datos es alta.